pdbfixer/hbonds.py

```diff
@@ -0,0 +1,293 @@
+            axis, atoms = residueRotations[residue]
+            center = positions[axis[1].index]
+            e = center-positions[axis[0].index]
+            e /= unit.norm(e)
+            for atom in atoms:
+                d = positions[atom.index]-center
+                dot = np.dot(e, d)
+                rotatedPositions[atom.index] = center - d + 2*dot*e
+
+        # For each variable residue, consider hydrogen bonds to other residues that are
+        # *not* variable.  How does the number change?
+
+        residueRotationHbondChange = defaultdict(int)
+        for residue in self.residueNeighbors:
+            for neighbor in self.residueNeighbors[residue]:
+                if neighbor not in self.residueNeighbors:
+                    before = self.countHbonds(residue, neighbor, positions, positions)
+                    after = self.countHbonds(residue, neighbor, rotatedPositions, positions)
+                    residueRotationHbondChange[residue] += after-before
+
+        # Now consider pairs of residues that are close enough to form hydrogen bonds to
+        # each other.
+
+        pairRotationHbonds = {}
+        for residue in self.residueNeighbors:
+            for neighbor in self.residueNeighbors[residue]:
+                if neighbor.index > residue.index and neighbor in self.residueNeighbors:
+                    pairRotationHbonds[(residue, neighbor)] = [
+                        [
+                            self.countHbonds(residue, neighbor, positions, positions),
+                            self.countHbonds(residue, neighbor, positions, rotatedPositions)
+                        ],
+                        [
+                            self.countHbonds(residue, neighbor, rotatedPositions, positions),
+                            self.countHbonds(residue, neighbor, rotatedPositions, rotatedPositions)
+                        ]
+                    ]
+
+        # Optimize each cluster.
+
+        newPositions = positions[:]
+        for cluster in clusters:
+            bestCombination = None
+            bestBonds = -1
+
+            # Loop over all possible combinations of which residues to rotate.
+
+            for combination in itertools.product(*[[0,1]]*len(cluster)):
+                bonds = 0
+                for rotated, residue in zip(combination, cluster):
+                    if rotated == 1:
+                        bonds += residueRotationHbondChange[residue]
+                    for neighbor in self.residueNeighbors[residue]:
+                        if (residue, neighbor) in pairRotationHbonds:
+                            neighborRotated = combination[cluster.index(neighbor)]
+                            bonds += pairRotationHbonds[(residue, neighbor)][rotated][neighborRotated]
+                if bonds > bestBonds:
+                    bestCombination = combination
+                    bestBonds = bonds
+
+            # Set the positions to the optimal set of rotations.
```

Comment:
Do we check that rotations do not cause clashes? The attached example is from PDB code 5UEX (H437 and V439). The histidine sidechain rotates to put two protons in close proximity. But, this code would probably capture this as two H-bonds and favor it (histidine N-H > backbone N and backbone N-H > histidine N). I would expect the backbone N-H to hydrogen bond with the histidine sidechain.
![prot_clash](https://github.com/user-attachments/assets/5a825756-579e-40c3-92e7-0fdaf8dd6028)

Reply:
It doesn't check for clashes, because most of them will be resolved by energy minimization.  In the image above, I don't see what the clash is?  The hydrogen pointing off to the left looks like it's out of the way of the other atoms.  It wouldn't be perceived as a second hydrogen bond because it isn't in line.  The angle between the D-H and H-A vectors can't be more than 60 degrees.  Also, the H-A distance has to be under 2.5 Å, and in that image it's 3.1 Å.

Reply:
Added a bit more context to the image this time. The backbone N-H is distorted (120 degrees). When minimized to ~180 degrees, it will be placed directly in line with the histidine sidechain N-H. Also when minimized, it could form a hydrogen bond to the histidine (if it was in the other protonation state).
![prot_clash2](https://github.com/user-attachments/assets/83d2ec3f-235d-44a2-b076-52df7417b3c7)

Reply:
I don't think I understand what you're saying.  If it rotates by 120 degrees as shown in the image, it will be pointing directly away from the HIS.  That will neither clash nor form a hydrogen bond.

What exactly is the issue you're describing, not just in this case but in general?  And what are you suggesting it should do differently?